File: zip2text_app/pipeline/job_manager.py
"""
Manages the orchestration of the entire OCR processing pipeline.

This module is the main entry point for a new processing job. It calls each of
the individual pipeline components in the correct order, handles errors that
may occur during the process, and ensures that cleanup is performed.
"""

import logging
import os
import shutil

# Import pipeline components
from .zip_handler import handle_zip_file
from .image_processor import scan_for_images
from .vision_client import perform_ocr_on_images
from .text_aggregator import aggregate_text_results

# Import real-time components
from ..realtime.event_streamer import EventStreamer
from ..realtime.log_formatter import Severity

def run_job(zip_file_path: str, job_id: str, streamer: EventStreamer) -> str:
    """
    Orchestrates the entire Zip-to-Text pipeline, emitting real-time events.

    This function defines the main sequence of operations for an OCR job:
    1. Emit a 'JOB_STARTED' event.
    2. Unzip the file using `handle_zip_file`.
    3. Scan for images using `scan_for_images`.
    4. Perform OCR on images using `perform_ocr_on_images`.
    5. Aggregate results with `aggregate_text_results`.
    6. Emit a 'JOB_COMPLETED' or 'JOB_FAILED' event.
    It also ensures that the temporary directory is cleaned up.

    Args:
        zip_file_path: The path to the uploaded .zip file.
        job_id: A unique identifier for this job.
        streamer: An initialized EventStreamer instance.

    Returns:
        The final aggregated text or an error message.
    """
    if not streamer:
        logging.error(f"[{job_id}] Event streamer was not provided to run_job!")
        # We can't emit a failure event here, so we just return. The job will hang.
        return "Error: Server configuration issue."

    streamer.emit_event(
        job_id=job_id,
        event_name='JOB_STARTED',
        status='RUNNING',
        severity=Severity.INFO,
        message=f"Processing new job for file: {os.path.basename(zip_file_path)}"
    )

    extracted_dir = None
    try:
        # Pass job_id and streamer to each pipeline step
        extracted_dir = handle_zip_file(zip_file_path, job_id, streamer)

        image_paths, image_count = scan_for_images(extracted_dir, job_id, streamer)

        if image_count == 0:
            message = "Process complete. No supported image files (.jpg, .png, .webp) were found."
            streamer.emit_event(
                job_id=job_id,
                event_name='JOB_WARNING',
                status='COMPLETED',
                severity=Severity.WARNING,
                message=message
            )
            return message

        ocr_results = perform_ocr_on_images(image_paths, job_id, streamer)

        final_text = aggregate_text_results(image_paths, ocr_results, job_id, streamer)

        streamer.emit_event(
            job_id=job_id,
            event_name='JOB_COMPLETED',
            status='SUCCESS',
            severity=Severity.SUCCESS,
            message="Successfully processed all images.",
            data={'final_text': final_text}
        )
        return final_text

    except ValueError as e:
        error_message = f"A validation error occurred: {e}"
        streamer.emit_event(
            job_id=job_id, event_name='JOB_FAILED', status='FAILED',
            severity=Severity.ERROR, message=error_message
        )
        return f"Error: {e}"
    except Exception as e:
        error_message = f"An unexpected error occurred: {e}"
        logging.error(f"[{job_id}] {error_message}", exc_info=True)
        streamer.emit_event(
            job_id=job_id, event_name='JOB_FAILED', status='FAILED',
            severity=Severity.ERROR, message="An unexpected server error occurred."
        )
        return "An unexpected server error occurred. Please try again."
    finally:
        if extracted_dir and os.path.exists(extracted_dir):
            try:
                shutil.rmtree(extracted_dir)
                logging.info(f"[{job_id}] Cleaned up temporary directory {extracted_dir}")
            except Exception as e:
                logging.error(f"[{job_id}] Failed to cleanup directory {extracted_dir}: {e}")

File: src/JobManager.php
<?php

namespace Zip2Text;

// This will require all the other pipeline components.
require_once __DIR__ . '/EventLogger.php';
require_once __DIR__ . '/ZipHandler.php';
require_once __DIR__ . '/ImageProcessor.php';
require_once __DIR__ . '/VisionClient.php';
require_once __DIR__ . '/TextAggregator.php';

class JobManager
{
    private array $job_data;
    private EventLogger $logger;
    private ?VisionClient $vision_client_override;

    public function __construct(array $job_data, ?VisionClient $vision_client_override = null)
    {
        $this->job_data = $job_data;
        $this->logger = new EventLogger($job_data['job_id']);
        $this->vision_client_override = $vision_client_override;
    }

    public function run(): void
    {
        $job_id = $this->job_data['job_id'];
        $zip_file_path = $this->job_data['zip_file_path'];
        $extracted_dir = null;

        try {
            $this->logger->emit('JOB_STARTED', 'INFO', 'Processing new job for file: ' . $this->job_data['original_filename']);

            // 1. Handle Zip File
            $zip_handler = new ZipHandler($job_id, $this->logger);
            $extracted_dir = $zip_handler->handle($zip_file_path);

            // 2. Scan for Images
            $image_processor = new ImageProcessor($job_id, $this->logger);
            $image_paths = $image_processor->scan($extracted_dir);

            if (empty($image_paths)) {
                $this->logger->emit('JOB_WARNING', 'WARNING', 'Process complete. No supported image files (.jpg, .png, .webp) were found.');
                // If no images, the job is done but not a failure.
                $this->cleanup($extracted_dir);
                return;
            }

            // 3. Perform OCR
            // Use the override if it exists, otherwise create a new client.
            $vision_client = $this->vision_client_override ?? new VisionClient($job_id, $this->logger);
            $ocr_results = $vision_client->performOcr($image_paths);

            // 4. Aggregate Text
            $text_aggregator = new TextAggregator($job_id, $this->logger);
            $final_text = $text_aggregator->aggregate($ocr_results);

            // 5. Job Completed
            $this->logger->emit('JOB_COMPLETED', 'SUCCESS', 'Successfully processed all images.', [
                'final_text' => $final_text
            ]);

        } catch (\Exception $e) {
            $this->logger->emit('JOB_FAILED', 'ERROR', 'An unexpected error occurred: ' . $e->getMessage());

        } finally {
            // 6. Cleanup
            $this->cleanup($extracted_dir);
        }
    }

    private function cleanup(?string $extracted_dir): void
    {
        // Clean up the temporary extracted directory
        if ($extracted_dir && is_dir($extracted_dir)) {
            // A recursive delete function is needed here.
            $this->rmdir_recursive($extracted_dir);
        }
    }

    /**
     * Recursively deletes a directory and its contents.
     */
    private function rmdir_recursive(string $dir): void
    {
        if (!is_dir($dir)) {
            return;
        }

        $items = new \RecursiveIteratorIterator(
            new \RecursiveDirectoryIterator($dir, \RecursiveDirectoryIterator::SKIP_DOTS),
            \RecursiveIteratorIterator::CHILD_FIRST
        );

        foreach ($items as $item) {
            if ($item->isDir()) {
                rmdir($item->getRealPath());
            } else {
                unlink($item->getRealPath());
            }
        }

        rmdir($dir);
    }
}
